### PR TITLE
Ensure global attributes added by intake-esm are compatible with netCDF and Zarr

### LINKED ISSUE #324: zarr does not like attrs added by intake-esm
Intake-esm adds the attribute `intake_esm_varname` to datasets, and I have encountered cases where that ends up being `None` (still looking for the exact model). 

Zarr does not like that type of metadata:
```python
import xarray as xr
ds_test = xr.DataArray(5).to_dataset(name='test')
ds_test.attrs['test'] = None

ds_test.to_zarr('test.zarr')
```

gives 

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-12-54ff2a1799bc> in <module>
      3 ds_test.attrs['test'] = None
      4 
----> 5 ds_test.to_zarr('test.zarr')

/srv/conda/envs/notebook/lib/python3.8/site-packages/xarray/core/dataset.py in to_zarr(self, store, chunk_store, mode, synchronizer, group, encoding, compute, consolidated, append_dim, region)
   1743             encoding = {}
   1744 
-> 1745         return to_zarr(
   1746             self,
   1747             store=store,

/srv/conda/envs/notebook/lib/python3.8/site-packages/xarray/backends/api.py in to_zarr(dataset, store, chunk_store, mode, synchronizer, group, encoding, compute, consolidated, append_dim, region)
   1453     # validate Dataset keys, DataArray names, and attr keys/values
   1454     _validate_dataset_names(dataset)
-> 1455     _validate_attrs(dataset)
   1456 
   1457     if mode == "a":

/srv/conda/envs/notebook/lib/python3.8/site-packages/xarray/backends/api.py in _validate_attrs(dataset)
    237     # Check attrs on the dataset itself
    238     for k, v in dataset.attrs.items():
--> 239         check_attr(k, v)
    240 
    241     # Check attrs on each variable within the dataset

/srv/conda/envs/notebook/lib/python3.8/site-packages/xarray/backends/api.py in check_attr(name, value)
    228 
    229         if not isinstance(value, (str, Number, np.ndarray, np.number, list, tuple)):
--> 230             raise TypeError(
    231                 f"Invalid value for attr {name!r}: {value!r} must be a number, "
    232                 "a string, an ndarray or a list/tuple of "

TypeError: Invalid value for attr 'test': None must be a number, a string, an ndarray or a list/tuple of numbers/strings for serialization to netCDF files
```

Should the attribute be set to a string `'none'` instead?

### LINKED ISSUE #474: Only add attributes if they have a value associated to them
**Is your feature request related to a problem? Please describe.**
I'm using a catalog for multiple dataset types (climate simulations, reanalyses, etc.). This means that for most models, not all columns will have a value associated to it. For example, ERA5 has no 'experiment_id', CMIP6 data has no 'driving_model_id', etc.

When I open a dataset using `.to_dataset_dict()` it will copy-paste the entries of all columns regardless, which means that I end up with many 'empty' attributes being added to the xarray object.

**Describe the solution you'd like**

`_open_dataset` in `source.py` has this:
```
            datasets = [
                _open_dataset(
                    record[self.path_column_name],
                    record[self.variable_column_name],
                    xarray_open_kwargs=_get_xarray_open_kwargs(
                        record['_data_format_'], self.xarray_open_kwargs, self.storage_options
                    ),
                    preprocess=self.preprocess,
                    expand_dims={
                        agg.attribute_name: [record[agg.attribute_name]]
                        for agg in self.aggregations
                        if agg.type.value == 'join_new'
                    },
                    requested_variables=self.requested_variables,
                    additional_attrs=record.to_dict(),
                )
                for _, record in self.df.iterrows()
            ]
```

A potential fix (that could be added as an optional argument) would be to have `additional_attrs=record[~record.isnull()].to_dict()` instead.

Would this be something that fits in `intake-esm` ?

### LINKED ISSUE #490: Attributes added by intake-esm contain invalid characters for netCDF
intake-esm creates global attributes based on the catalog. By default, they follow the format: intake_esm_attr/(x). While the first part of the "path" is customizable (#460), the separator character "/" is not.  However, it is an illegal character when writing netCDF files (it is valid for zarr datasets though).

Would it be reasonable to have this character:
- changed to something legal ("_", "-", "\", ...)
- be customizable by downstream applications
?

### PR DESCRIPTION
**Pull request by andersy005**
## Related issue number

- Fixes #490 
- Fixes #324 
- Fixes #474

## Checklist

- [x] Unit tests for the changes exist
- [x] Tests pass on CI
- [x] Documentation reflects the changes where applicable

### CODE REVIEW

**Comment by andersy005** [comment]
@aulemahal, when you get a moment, can you take a look at this and let me know if you have any feedback?

File: intake_esm/source.py [hunk 1]
```diff
@@ -229,7 +232,7 @@ def _open_dataset(self):
                     },
                     requested_variables=self.requested_variables,
                     data_format=record['_data_format_'],
-                    additional_attrs=record.to_dict(),
+                    additional_attrs=record[~record.isnull()].to_dict(),
```
Comment (mgrover1) [inline intake_esm/source.py:235]: I like this approach..

**Review by mgrover1 — APPROVED** [review]
Great! Thanks for putting this together @andersy005

**Review by aulemahal — APPROVED** [review]
Nice! Thanks for solving these issues!

**Comment by andersy005** [comment]
thank you for the feedback , @aulemahal & @mgrover1

**Pull request merged**